diff --git a/packages/hydrooj/src/handler/problem.ts b/packages/hydrooj/src/handler/problem.ts
index e2fa80ed..c45ead6a 100644
--- a/packages/hydrooj/src/handler/problem.ts
+++ b/packages/hydrooj/src/handler/problem.ts
@@ -352,6 +352,7 @@ export class ProblemDetailHandler extends ContestDetailBaseHandler {
             discussion.count(domainId, { parentId: this.pdoc.docId }),
         ]);
         this.response.body = {
+	    callfrom: 'myself',
             pdoc: this.pdoc,
             udoc: this.udoc,
             psdoc: tid ? null : this.psdoc,
@@ -458,6 +459,24 @@ export class ProblemDetailHandler extends ContestDetailBaseHandler {
         this.back({ star });
     }
 }
+export class ProblemSlidevHandler extends ProblemDetailHandler {
+    @param('tid', Types.ObjectId, true)
+    async prepare(domainId: string, tid?: ObjectId) {
+        if (tid && !contest.isOngoing(this.tdoc, this.tsdoc)) throw new ContestNotLiveError(this.tdoc.docId);
+        if (typeof this.pdoc.config === 'string') throw new ProblemConfigError();
+        if (this.pdoc.config.langs && !this.pdoc.config.langs.length) throw new ProblemConfigError();
+    }
+
+    async get() {
+        this.response.template = 'problem_detail.html';
+        const langRange = (typeof this.pdoc.config === 'object' && this.pdoc.config.langs)
+            ? Object.fromEntries(this.pdoc.config.langs.map((i) => [i, setting.langs[i]?.display || i]))
+            : setting.SETTINGS_BY_KEY.codeLang.range;
+        this.response.body.callfrom = 'slidev';
+        this.response.body.page_name = 'problem_detail';
+    }
+}
+
 
 export class ProblemSubmitHandler extends ProblemDetailHandler {
     @param('tid', Types.ObjectId, true)
@@ -1062,6 +1081,7 @@ export async function apply(ctx: Context) {
     ctx.Route('problem_main', '/p', ProblemMainHandler, PERM.PERM_VIEW_PROBLEM);
     ctx.Route('problem_random', '/problem/random', ProblemRandomHandler, PERM.PERM_VIEW_PROBLEM);
     ctx.Route('problem_detail', '/p/:pid', ProblemDetailHandler);
+    ctx.Route('problem_detail', '/p/:pid/slidev', ProblemSlidevHandler);
     ctx.Route('problem_submit', '/p/:pid/submit', ProblemSubmitHandler, PERM.PERM_SUBMIT_PROBLEM);
     ctx.Route('problem_hack', '/p/:pid/hack/:rid', ProblemHackHandler, PERM.PERM_SUBMIT_PROBLEM);
     ctx.Route('problem_edit', '/p/:pid/edit', ProblemEditHandler);
